common.py

```diff
@@ -0,0 +1,243 @@
+    if references_dir is None:
+        raise ValueError('References dir not set')
+    return references_dir
+
+
+def references_dict(config):
+    """
+    The config file is designed to be easy to edit and use from the user's
+    standpoint. But it's not so great for practical usage. Here we convert the
+    config file which has the format::
+
+    >>> from textwrap import dedent
+    >>> fout = open('tmp', 'w')
+    >>> _ = fout.write(dedent('''
+    ... references_dir: "/data"
+    ... references:
+    ...   dm6:
+    ...     r6-11:
+    ...       fasta:
+    ...         url: ""
+    ...         indexes:
+    ...           - bowtie2
+    ...           - hisat2
+    ...       gtf:
+    ...         url: ""
+    ...         conversions:
+    ...           - refflat
+    ...     r6-11_transcriptome:
+    ...       fasta:
+    ...         url: ""
+    ...         indexes:
+    ...           - kallisto
+    ... '''))
+    >>> fout.close()
+
+    To this format:
+
+    >>> d = config_to_dict('tmp')
+    >>> assert d == (
+    ... {
+    ...   'dm6': {
+    ...      'r6-11': {
+    ...          'fasta': '/data/dm6/fasta/dm6_r6-11.fasta',
+    ...          'refflat': '/data/dm6/gtf/dm6_r6-11.refflat',
+    ...          'gtf': '/data/dm6/gtf/dm6_r6-11.gtf',
+    ...          'chromsizes': '/data/dm6/fasta/dm6_r6-11.chromsizes',
+    ...          'bowtie2': '/data/dm6/bowtie2/dm6_r6-11.1.bt2',
+    ...          'hisat2': '/data/dm6/hisat2/dm6_r6-11.1.ht2',
+    ...          },
+    ...      'r6-11_transcriptome': {
+    ...          'fasta': '/data/dm6/fasta/dm6_r6-11_transcriptome.fasta',
+    ...          'chromsizes': '/data/dm6/fasta/dm6_r6-11_transcriptome.chromsizes',
+    ...          'kallisto': '/data/dm6/kallisto/dm6_r6-11_transcriptome.idx',
+    ...          },
+    ...     },
+    ... }), d
+    >>> os.unlink('tmp')
+
+    With this new dictionary, other parts of the config or snakemake rules can
+    access the files by d[assembly][tag][type].
+    """
+    if isinstance(config, str):
+        config = yaml.load(open(config))
+
+    references_dir = get_references_dir(config)
+
+    # Map "indexes" value to a pattern specific to each index.
+    index_extensions = {
+        'bowtie2': aligners.bowtie2_index_from_prefix('')[0],
+        'hisat2': aligners.hisat2_index_from_prefix('')[0],
+        'kallisto': '.idx',
+    }
+
+
+    conversion_extensions = {
+        'intergenic': '.intergenic.gtf',
+        'refflat': '.refflat',
+    }
+
+    d = {}
+    for assembly in config['references'].keys():
+        d[assembly] = {}
+        for tag in config['references'][assembly].keys():
+            e = {}
+            for type_, block in config['references'][assembly][tag].items():
+                e[type_] = (
+                    '{references_dir}/'
+                    '{assembly}/'
+                    '{type_}/'
+                    '{assembly}_{tag}.{type_}'.format(**locals())
+                )
```

Comment:
Wow, this is super creative use of locals(). I don't think I would have ever come up with that.